src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java

```diff
@@ -287,24 +279,46 @@ private static String formatMessageBatchToStore(GroupedMessageBatchToStore batch
      */
     private void verifyBatch(T delivery) {
         List<M> messages = getMessages(delivery);
+        HashMap<SessionKey, SessionData> innerCache = new HashMap<>();
         SessionKey lastKey = null;
-        long lastSeq = Long.MIN_VALUE;
+        long prevSequence = Long.MIN_VALUE;
         for (int i = 0; i < messages.size(); i++) {
             M message = messages.get(i);
             SessionKey sessionKey = createSessionKey(message);
+            long currentSeq = extractSequence(message);
+
+            if (innerCache.containsKey(sessionKey)) {
+                prevSequence = innerCache.get(sessionKey).lastSequence.get();
+            } else if(sessionData.containsKey(sessionKey)) {
+                prevSequence = sessionData.get(sessionKey).lastSequence.get();
+            } else {
+                prevSequence = getLastMessageSequence(sessionKey);
+            }
+
             if (lastKey == null) {
                 lastKey = sessionKey;
             } else {
-                verifySession(i, lastKey, sessionKey);
+                verifySessionGroup(i, lastKey, sessionKey);
```

Comment:
is it ever checked that messages of same session across batches have same group?

Reply:
in verifySessionGroup method groups of lastKey and sessionKey are compared and it throws exception if they are different